<a href="https://colab.research.google.com/github/stuck-w-yu/Iris_dataset/blob/main/Pert_2_MachineLearninf.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Buat model machine learning menggunakan google colab untuk memprediksi harga rumah menggunakan dataset "/content/Rumah.csv". Lakukan data preprocessing, latih model regresi linear dan regresi polinomial, lalu evaluasi dan bandingkan kedua model tersebut.

## Memuat data

### Subtask:
Memuat dataset harga rumah dari file CSV ke dalam DataFrame pandas.


**Reasoning**:
Import the pandas library and load the dataset into a DataFrame, then display the first 5 rows.



In [ ]:
import pandas as pd

df = pd.read_csv("/content/Rumah.csv")
display(df.head())

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

## Preprocessing data

### Subtask:
Membersihkan dan mempersiapkan data untuk pelatihan model, termasuk menangani nilai yang hilang (jika ada), mengkodekan variabel kategori (jika ada), dan membagi data menjadi set pelatihan dan pengujian.


**Reasoning**:
Periksa apakah ada nilai yang hilang dalam DataFrame `df` menggunakan metode `.isnull().sum()`.



In [ ]:
print(df.isnull().sum())

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64


**Reasoning**:
Tidak ada nilai yang hilang dalam dataset, dan tidak ada variabel kategorikal yang perlu diencode. Langkah selanjutnya adalah memisahkan fitur dan target, lalu membagi data menjadi set pelatihan dan pengujian.



In [ ]:
from sklearn.model_selection import train_test_split

X = df.drop('MEDV', axis=1)
y = df['MEDV']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (404, 13)
Shape of X_test: (102, 13)
Shape of y_train: (404,)
Shape of y_test: (102,)


## Melatih model regresi linear

### Subtask:
Melatih model regresi linear menggunakan data pelatihan untuk memprediksi harga rumah berdasarkan fitur yang relevan.


**Reasoning**:
Import the LinearRegression class, create an instance of the model, and train it using the training data.



In [ ]:
from sklearn.linear_model import LinearRegression

linear_model = LinearRegression()
linear_model.fit(X_train, y_train)

LinearRegression()

## Evaluasi model regresi linear

### Subtask:
Mengevaluasi kinerja model regresi linear menggunakan data pengujian dan metrik yang sesuai (misalnya, R-squared, Mean Squared Error).


**Reasoning**:
Evaluate the linear regression model performance using Mean Squared Error and R-squared on the test set.



In [ ]:
from sklearn.metrics import mean_squared_error, r2_score

y_pred_linear = linear_model.predict(X_test)

mse_linear = mean_squared_error(y_test, y_pred_linear)
r2_linear = r2_score(y_test, y_pred_linear)

print(f"Linear Regression Model - Mean Squared Error: {mse_linear:.2f}")
print(f"Linear Regression Model - R-squared: {r2_linear:.2f}")

Linear Regression Model - Mean Squared Error: 24.29
Linear Regression Model - R-squared: 0.67


## Melatih model regresi polinomial

### Subtask:
Melatih model regresi polinomial menggunakan data pelatihan untuk memprediksi harga rumah berdasarkan fitur yang relevan. Ini mungkin melibatkan pembuatan fitur polinomial baru.


**Reasoning**:
Import necessary libraries, create polynomial features for the training data, and train a linear regression model on the transformed data.



In [ ]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression

poly_features = PolynomialFeatures(degree=2)
X_train_poly = poly_features.fit_transform(X_train)

poly_model = LinearRegression()
poly_model.fit(X_train_poly, y_train)

X_test_poly = poly_features.transform(X_test)

## Evaluasi model regresi polinomial

### Subtask:
Mengevaluasi kinerja model regresi polinomial menggunakan data pengujian dan metrik yang sesuai.


**Reasoning**:
Calculate the predicted values for the polynomial model on the test set, then calculate and print the MSE and R-squared scores.



In [ ]:
from sklearn.metrics import mean_squared_error, r2_score

y_pred_poly = poly_model.predict(X_test_poly)

mse_poly = mean_squared_error(y_test, y_pred_poly)
r2_poly = r2_score(y_test, y_pred_poly)

print(f"Polynomial Regression Model - Mean Squared Error: {mse_poly:.2f}")
print(f"Polynomial Regression Model - R-squared: {r2_poly:.2f}")

Polynomial Regression Model - Mean Squared Error: 14.26
Polynomial Regression Model - R-squared: 0.81


## Membandingkan model

### Subtask:
Membandingkan kinerja model regresi linear dan regresi polinomial untuk menentukan model mana yang memberikan prediksi terbaik.


**Reasoning**:
Display the previously calculated MSE and R2 values for both models and interpret them to determine the better performing model.



In [ ]:
print("--- Model Comparison ---")
print(f"Linear Regression Model - Mean Squared Error: {mse_linear:.2f}")
print(f"Linear Regression Model - R-squared: {r2_linear:.2f}")
print(f"Polynomial Regression Model - Mean Squared Error: {mse_poly:.2f}")
print(f"Polynomial Regression Model - R-squared: {r2_poly:.2f}")

if mse_poly < mse_linear and r2_poly > r2_linear:
    print("\nConclusion: The Polynomial Regression Model performed better.")
elif mse_linear < mse_poly and r2_linear > r2_poly:
    print("\nConclusion: The Linear Regression Model performed better.")
else:
    print("\nConclusion: The performance of both models is comparable.")

--- Model Comparison ---
Linear Regression Model - Mean Squared Error: 24.29
Linear Regression Model - R-squared: 0.67
Polynomial Regression Model - Mean Squared Error: 14.26
Polynomial Regression Model - R-squared: 0.81

Conclusion: The Polynomial Regression Model performed better.


## Summary:

### Data Analysis Key Findings

*   The dataset contains no missing values and consists entirely of numerical variables.
*   The data was successfully split into training and testing sets (80% train, 20% test).
*   The Linear Regression model achieved a Mean Squared Error (MSE) of 24.29 and an R-squared score of 0.67 on the test data.
*   The Polynomial Regression model (with degree 2) achieved a Mean Squared Error (MSE) of 14.26 and an R-squared score of 0.81 on the test data.

### Insights or Next Steps

*   The Polynomial Regression model performed better than the Linear Regression model based on the lower MSE and higher R-squared score, suggesting that a non-linear relationship exists between the features and house prices.
*   Further investigation could involve exploring different polynomial degrees or considering other regression models to potentially improve prediction accuracy.
